hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/SnapshotDeletingService.java

```diff
@@ -0,0 +1,378 @@
+          <String, SnapshotInfo>> iterator = snapshotInfoTable.iterator()) {
+        iterator.seekToFirst();
+
+        long snapshotLimit = snapshotDeletionPerTask;
+
+        while (iterator.hasNext() && snapshotLimit > 0) {
+          SnapshotInfo snapInfo = iterator.next().getValue();
+          SnapshotInfo.SnapshotStatus snapshotStatus =
+              snapInfo.getSnapshotStatus();
+
+          // Only Iterate in deleted snapshot
+          if (!snapshotStatus.equals(
+              SnapshotInfo.SnapshotStatus.SNAPSHOT_DELETED)) {
+            continue;
+          }
+
+          OmSnapshot omSnapshot = (OmSnapshot) omSnapshotManager
+              .checkForSnapshot(snapInfo.getVolumeName(),
+                  snapInfo.getBucketName(),
+                  getSnapshotPrefix(snapInfo.getName()));
+
+          Table<String, RepeatedOmKeyInfo> snapshotDeletedTable =
+              omSnapshot.getMetadataManager().getDeletedTable();
+
+          if (snapshotDeletedTable.getEstimatedKeyCount() == 0) {
+            continue;
+          }
+
+          // Get bucketInfo for the snapshot bucket to get bucket layout.
+          String dbBucketKey = ozoneManager.getMetadataManager().getBucketKey(
+              snapInfo.getVolumeName(), snapInfo.getBucketName());
+          OmBucketInfo bucketInfo = ozoneManager.getMetadataManager()
+              .getBucketTable().get(dbBucketKey);
+
+
+          chainManager.loadFromSnapshotInfoTable(
+              ozoneManager.getMetadataManager());
+          SnapshotInfo nextSnapshot = getNextNonDeletedSnapshot(snapInfo);
+
+          SnapshotInfo previousSnapshot = getPreviousSnapshot(snapInfo);
+          Table<String, OmKeyInfo> previousKeyTable = null;
+          OmSnapshot omPreviousSnapshot = null;
+
+          // Handle case when the deleted snapshot is the first snapshot.
+          // Move deleted keys to activeDB's deletedKeyTable
+          if (previousSnapshot != null) {
+            omPreviousSnapshot = (OmSnapshot) omSnapshotManager
+                .checkForSnapshot(previousSnapshot.getVolumeName(),
+                    previousSnapshot.getBucketName(),
+                    getSnapshotPrefix(previousSnapshot.getName()));
+
+            previousKeyTable = omPreviousSnapshot
+                .getMetadataManager().getKeyTable(bucketInfo.getBucketLayout());
+          }
+
+          // Move key to either next non deleted snapshot's snapshotDeletedTable
+          // or move to active object store deleted table
+
+          List<KeyValuePair> toActiveDBList = new ArrayList<>();
+          List<KeyValuePair> toNextDBList = new ArrayList<>();
+
+          try (TableIterator<String, ? extends Table.KeyValue<String,
+              RepeatedOmKeyInfo>> deletedIterator = snapshotDeletedTable
+              .iterator()) {
+
+            iterator.seekToFirst();
+
+            while (deletedIterator.hasNext()) {
+              Table.KeyValue<String, RepeatedOmKeyInfo>
```

Comment:
Do we need set limit for number of deletedKeys? similar to keyDeletingService

Reply:
@sumitagrawl Yes we do need that. That can be made a configurable value.

Reply:
Yes we can do that.

Reply:
Finishing this up here [HDDS-7883](https://issues.apache.org/jira/browse/HDDS-7883)